examples/plot_metadata_routing.py

```diff
@@ -0,0 +1,624 @@
+
+# %%
+# In the above example, this is how each ``fit`` method will call the
+# sub-estimator's ``fit``:
+#
+#     meta_est.fit(X, y, aliased_foo=my_weights):
+#         ...  # this estimator (est), expects aliased_foo as seen above
+#         self.estimator_.fit(X, y, aliased_foo=aliased_foo):
+#             ...  # est passes aliased_foo's value as foo, which is expected
+#                  # by the sub-estimator
+#             self.estimator_.fit(X, y, foo=aliased_foo)
+#    ...
+
+# %%
+# Router and Consumer
+# -------------------
+# To show how a slightly more complicated case would work, consider a case
+# where a meta-estimator uses some metadata, but it also routes them to an
+# underlying estimator. In this case, this meta-estimator is a consumer and a
+# router at the same time. This is how we can implement one, and it is very
+# similar to what we had before, with a few tweaks.
+
+
+class RouterConsumerClassifier(MetaEstimatorMixin, ClassifierMixin, BaseEstimator):
+    def __init__(self, estimator, foo_is_none=True):
+        self.estimator = estimator
+        self.foo_is_none = foo_is_none
+
+    def fit(self, X, y, foo, **fit_params):
+        if self.estimator is None:
+            raise ValueError("estimator cannot be None!")
+
+        if (foo is None) != self.foo_is_none:
+            raise ValueError("foo's value and foo_is_none disagree!")
+
+        if foo is not None:
+            fit_params["foo"] = foo
+
+        # meta-estimators are responsible for validating the given metadata
+        metadata_request_factory(self).fit.validate_metadata(
+            ignore_extras=False, self_metadata=super(), kwargs=fit_params
+        )
+        # we can use provided utility methods to map the given metadata to what
+        # is required by the underlying estimator
+        fit_params_ = metadata_request_factory(self.estimator).fit.get_method_input(
+            ignore_extras=False, kwargs=fit_params
+        )
+        self.estimator_ = clone(self.estimator).fit(X, y, **fit_params_)
+        self.classes_ = self.estimator_.classes_
+        return self
+
+    def predict(self, X, **predict_params):
+        check_is_fitted(self)
+        # same as in ``fit``, we validate the given metadata
+        metadata_request_factory(self).predict.validate_metadata(
+            ignore_extras=False, kwargs=predict_params
+        )
+        # and then prepare the input to the underlying ``predict`` method.
+        predict_params_ = metadata_request_factory(
+            self.estimator_
+        ).predict.get_method_input(ignore_extras=False, kwargs=predict_params)
+        return self.estimator_.predict(X, **predict_params_)
+
+    def get_metadata_request(self):
+        router = (
+            MetadataRouter()
+            .add(super(), mapping="one-to-one", overwrite=False, mask=False)
+            .add(self.estimator, mapping="one-to-one", overwrite="smart", mask=True)
```

Comment:
I didn't read the documentation yet but at this point `overwrite="smart"` is magical and smart :)

Reply:
The other PR now doesn't have overwrite and mask, should be much easier to review/understand.